gc/base/standard/Scavenger.cpp

```diff
@@ -479,14 +485,100 @@ MM_Scavenger::calculateMaxCacheCount(uintptr_t activeMemorySize)
+
+	uint64_t totalStallTime =  avgTimeToStartCollection + avgTimeIdleAfterCollection + avgScanStallTime + avgSyncStallTime +  avgNotifyStallTime;
+	uint64_t scavengeTotalTime = _extensions->scavengerStats._endTime - _extensions->scavengerStats._startTime;
+
+	totalStallTime = omrtime_hires_delta(0, totalStallTime, OMRPORT_TIME_DELTA_IN_MICROSECONDS);
+	scavengeTotalTime =	omrtime_hires_delta (0, scavengeTotalTime, OMRPORT_TIME_DELTA_IN_MICROSECONDS);
+
+
+	/* This Adaptive Threading Model aims to determine the efficiency of a cycle and predict the optimal GC thread count based on current number of threads,
+	 * directly proportional to busy time and inversely proportional to idle times.
+	 *
+	 * The model can be expressed as a continues function, it's derived by finding a minimum of the folowing GC time function (used to project duration of GC for m threads,
+	 * with oberved buys/stall times while performing GC with n threads):
+	 *
+	 *          Time GC (m,n,b,s) = b * (n/m) + s * (m/n)^x
+	 * Where m = number of threads for which total GC time (duration) is projected
+	 *       n = number of utilized threads (number of worker threads started + main thread) for the GC cycle used to observe s and b
+	 *       s = stall time per thread = average observed collection stall time for n threads = SUM(Stall time of n threads) / n
+	 *       b = busy time per thread = average collection busy time for n threads
+	 *       X = Stall Overhead Sensitivity/Tolerance, a model constant used to help model non-linear dependency of stall times on GC thread count
+	 *
+	 *
+	 * Solving this function results in the model implementation expressed by expression 1 below. Expression 1 in combination with expression 2 gives up a complete implementation of the model
+	 *
+	 *  (1) Number of Optimal Threads = m(n,b,s) = n * (B/X*s)^(1/X+1)
+	 *  (2) floor(((m(n,b,s) + H) * (1 - W)) + (n * W))
+	 *
+	 * Where W and H are constants, W = Weighted average factor, H = Thread Booster
+	 *
+	 * Expression (1) can be simplified to be written in terms of % stall and working threads as follows:
+	 *
+	 *                    b/s = (1/%stall) - 1
+	 *
+	 *  ------------------------------------------------------------------
+	 *  | (1) m(n,b,s) = m(n,%stall) = n * ((1/x)*(1/%stall - 1)^(1/(x+1))|
+	 *  -------------------------------------------------------------------
+	 */
+	double percentStall = ((double) totalStallTime) / ((double) scavengeTotalTime);
+	double modifier = _extensions->adaptiveThreadingSensitivityFactor;
+	double powerModifier = modifier + 1.0;
+	double stallModifider = (1.0 / percentStall) - 1.0;
+	double powerBase = (1.0 / modifier) * stallModifider;
+
+	double idealThreads = _dispatcher->activeThreadCount() * pow(powerBase, (1 / powerModifier));
+	float adjustedAverage = MM_Math::weightedAverage((float)_dispatcher->activeThreadCount(), (float)idealThreads, _extensions->adaptiveThreadingWeightActiveThreads);
```

Comment:
You can introduce new API for double to avoid casting. Although the whole preceding calculation and all params could've been floats, too.

Reply:
Fixed - changed all preceding calculations and params to float. With this changed, the extensions variables are changed to float as well.

Reply:
https://github.com/RSalman/omr/blob/893eb8f9413eb74eecdd8135ea4b62e115ab019f/gc/base/GCExtensionsBase.hpp#L1619-L1621

Reply:
https://github.com/RSalman/omr/blob/893eb8f9413eb74eecdd8135ea4b62e115ab019f/gc/base/GCExtensionsBase.hpp#L1619-L1621